e2e-testing/readme.md

```diff
@@ -0,0 +1,164 @@
+When working on large and complex customer engagements, it is very important to perform E2E testing in a controlled environment, and for a good reason. Even  slight change in the sub-system implementation, the interface contract, data schema, etc. can cause interface conflicts. Therefore, we need **constantly** and **automatically** control E2E tests to manage the scope and breadth of the conflicts.
+
+## E2E Testing Methods
+
+### Horizontal Test
+
+This method is used very commonly. It occurs horizontally across the context of multiple applications. Take an example of a web-based application of an online ordering system. The whole process will include accounts, inventory status of the products as well as shipping details.
+
+### Vertical Test
+
+In this method, all the transactions of any application are verified and evaluated right from the start to finish. Each individual layer of the application is tested starting from top to bottom. Take an example of a web-based application that uses middleware API for reaching back-end services. In such case, API is required to be fully tested against both, the underlying back-end services and the top layer UI web interface. All these complex testing scenarios will require proper validation and dedicated automated testing. Thus this method is much more difficult.
+
+Below enlisted are few **guidelines** that should be kept in mind while designing the test cases for performing E2E testing:
+
+- Test cases should be designed from the end user’s perspective.
+- Should focus on testing some existing features of the system.
+- Multiple scenarios should be considered for creating multiple test cases.
+- Different sets of test cases should be created to focus on multiple scenarios of the system.
+
+## E2E Testing Design Framework
+
+![E2E Testing Design Framework](./images/e2e-testing-design-framework.jpg)
+
+We will look into all the 3 categories one by one:
+
+### User Functions
+
+Following actions should be performed as a part of building user functions:
+
+- List user initiated functions of the software systems and their interconnected sub-systems.
+- For any function, keep track of the actions performed as well as Input and Output data.
+- Find the relations, if any between different Users functions.
+- Find out the nature of different user functions i.e. if they are independent or are reusable.
+
+### Conditions
+
+Following activities should be performed as a part of building conditions based on user functions:
+
+- For each and every user functions, a set of conditions should be prepared.
+- Timing, data conditions and other factors that affect user functions can be considered as parameters.
+
+### Test Cases
+
+Following factors should be considered for building test cases:
+
+- For every scenario, one or more test cases should be created to test each and every functionality of the user functions.
+- Every single condition should be enlisted as a separate test case.
+
+## E2E Testing Progress Tracking
```

Comment:
Can you talk about this being part of the CI/CD process and failing the build?

Reply:
added to the document